src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Snapshot/Filters/IAddDependencyContext.cs

```diff
@@ -0,0 +1,55 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Tree.Dependencies.Snapshot.Filters
+{
+    /// <summary>
+    /// Context used by <see cref="IDependenciesSnapshotFilter"/> implementations when filtering
+    /// a dependency that is being added to a <see cref="DependenciesSnapshot"/> by
+    /// <see cref="DependenciesSnapshot.FromChanges"/>.
+    /// </summary>
+    internal interface IAddDependencyContext
+    {
+        /// <summary>
+        /// Indicates the filter wishes to add <paramref name="dependency"/> to the snapshot.
+        /// </summary>
+        /// <remarks>
+        /// Note that this may not be the instance the filter was provided, as it may mutate it
+        /// in some way prior to adding it to the snapshot. Also note that when multiple filters
+        /// exist, later filters may further mutate or reject this filter's accepted dependency.
+        /// </remarks>
+        /// <param name="dependency">The dependency to add to the snapshot.</param>
+        void Accept(IDependency dependency);
+
+        /// <summary>
+        /// Indicates the filter rejects the dependency to be added to the snapshot.
+        /// </summary>
+        void Reject();
+
+        /// <summary>
+        /// Adds a new, or replaces an existing dependency (keyed on <see cref="IDependencyModel.Id"/>).
+        /// </summary>
+        /// <remarks>
+        /// In the course of filtering one dependency, the filter may wish to modify or add other
+        /// dependencies in the project's tree. This method allows that to happen.
+        /// </remarks>
+        void AddOrUpdate(IDependency dependency);
+
+        /// <summary>
+        /// Attempts to find the dependency in the project's tree with specified <paramref name="dependencyId"/>.
+        /// </summary>
+        bool TryGetDependency(string dependencyId, out IDependency dependency);
```

Comment:
Is there a reason you decided to expose these three methods, instead of just exposing the "world" as a property? You're already tied to an implementation via ImmutableDictionary? It is to avoid accidental boxing the of the enumerator?

Reply:
Ah, to avoid exposing the "builder"?